### ISSUE TITLE
Implementing BatchNorm with NEConvolution

### ISSUE THREAD

**Issue by sagarkar10** [ARM-software/ComputeLibrary#130]
I am trying to implement batchnormalization before the activation function and after the NEConvolutional.
I want to chose the batchnorm parameters (mean, variance, etc) from the weight files.
Would appriciate some help.

**Comment by GeorgeARM** [comment]
Hello @sagarkar10,

What issues are you facing?
By the way, batch normalization will be part of the next release (hopefully in a couple of days).

**Comment by Rockyzhang-DeepViewTech** [comment]
@GeorgeARM 
Will the next release also have the CL version BatchNorm?
That will be great, looking forward to it!

**Comment by GeorgeARM** [comment]
Yes @Rockyzhang-DeepViewTech for both NEON and CL.

**Comment by sagarkar10** [comment]
Hi @GeorgeARM I was trying to implement the darknet model for arm.
```C
 void forward_convolutional_layer(convolutional_layer l, network_state state)
{
    int out_h = l.out_h;
    int out_w = l.out_w;
    int i;

    fill_cpu(l.outputs*l.batch, 0, l.output, 1);

    if(l.xnor){
        binarize_weights(l.weights, l.n, l.c*l.size*l.size, l.binary_weights);
        swap_binary(&l);
        binarize_cpu(state.input, l.c*l.h*l.w*l.batch, l.binary_input);
        state.input = l.binary_input;
    }

    int m = l.n;
    int k = l.size*l.size*l.c;
    int n = out_h*out_w;


    float *a = l.weights;
    float *b = state.workspace;
    float *c = l.output;

    for(i = 0; i < l.batch; ++i){
        im2col_cpu(state.input, l.c, l.h, l.w, 
                l.size, l.stride, l.pad, b);
        gemm(0,0,m,n,k,1,a,k,b,n,1,c,n);
        c += n*m;
        state.input += l.c*l.h*l.w;
    }

if(l.batch_normalize){
        forward_batchnorm_layer(l, state);
} else {
        add_bias(l.output, l.biases, l.batch, l.n, out_h*out_w);
}

    activate_array(l.output, m*n*l.batch, l.activation);
    if(l.binary || l.xnor) swap_binary(&l);
}
```
I want some help implementing `forward_batchnorm_layer( l, state )`

**Comment by GeorgeARM** [comment]
Hello @sagarkar10,

Usually you compute mean and variance during training, thus your normalization ends up being a linear transformation. So you have to dump those from the model and fill your tensors (You can have a look how to fill a tensor in compute library documentation). You should dump the raw data of each trainable parameter to a separate file for now as we don't have an interface to parse weight files from other frameworks at the moment.
Once you read the data you just have to perform a simple transformation of each of the activations.
You can just loop over the input tensor and perform this.
As I mentioned above we have already implemented BatchNormalization and will be part of our next release in a couple of days.

**Comment by austingg** [comment]
@GeorgeARM @sagarkar10  mobile device is usually used for inference which the parameter of BN is fixed and can be merged into conv's param.

**Comment by GeorgeARM** [comment]
Closing this as BatchNormalizationLayer was added in the latest release(v17.06). Feel free to reopen if needed.
